src/coreclr/src/jit/hwintrinsiccodegenarm64.cpp

```diff
@@ -16,266 +16,244 @@
-    const bool tableDrivenCategory =
-        (category != HW_Category_Special) && (category != HW_Category_Scalar) && (category != HW_Category_Helper);
-    const bool tableDrivenFlag =
-        !HWIntrinsicInfo::GeneratesMultipleIns(intrinsicId) && !HWIntrinsicInfo::HasSpecialCodegen(intrinsicId);
-    return tableDrivenCategory && tableDrivenFlag;
-}
+    HWIntrinsic(GenTreeHWIntrinsic* node)
+        : op1(nullptr), op2(nullptr), op3(nullptr), numOperands(0), baseType(TYP_UNDEF)
+    {
+        assert(node != nullptr);
 
-//------------------------------------------------------------------------
-// genHWIntrinsic: Generates the code for a given hardware intrinsic node.
-//
-// Arguments:
-//    node - The hardware intrinsic node
-//
-void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
-{
-    NamedIntrinsic      intrinsicId = node->gtHWIntrinsicId;
-    HWIntrinsicCategory category    = HWIntrinsicInfo::lookupCategory(intrinsicId);
+        id       = node->gtHWIntrinsicId;
+        category = HWIntrinsicInfo::lookupCategory(id);
+
+        assert(HWIntrinsicInfo::RequiresCodegen(id));
 
-    assert(HWIntrinsicInfo::RequiresCodegen(intrinsicId));
+        InitializeOperands(node);
+        InitializeBaseType(node);
+    }
 
-    if (genIsTableDrivenHWIntrinsic(intrinsicId, category))
+    bool IsTableDriven() const
     {
-        InstructionSet isa     = HWIntrinsicInfo::lookupIsa(intrinsicId);
-        int            ival    = HWIntrinsicInfo::lookupIval(intrinsicId);
-        int            numArgs = HWIntrinsicInfo::lookupNumArgs(node);
+        // TODO-Arm64-Cleanup - make more categories to the table-driven framework
+        bool isTableDrivenCategory = (category != HW_Category_Special) && (category != HW_Category_Helper);
+        bool isTableDrivenFlag = !HWIntrinsicInfo::GeneratesMultipleIns(id) && !HWIntrinsicInfo::HasSpecialCodegen(id);
+
+        return isTableDrivenCategory && isTableDrivenFlag;
+    }
 
-        assert(numArgs >= 0);
+    NamedIntrinsic      id;
+    HWIntrinsicCategory category;
+    GenTree*            op1;
+    GenTree*            op2;
+    GenTree*            op3;
+    int                 numOperands;
+    var_types           baseType;
 
-        GenTree*  op1        = node->gtGetOp1();
-        GenTree*  op2        = node->gtGetOp2();
-        regNumber targetReg  = node->GetRegNum();
-        var_types targetType = node->TypeGet();
-        var_types baseType   = node->gtSIMDBaseType;
+private:
+    void InitializeOperands(GenTreeHWIntrinsic* node)
+    {
+        op1 = node->gtGetOp1();
+        op2 = node->gtGetOp2();
 
-        instruction ins = HWIntrinsicInfo::lookupIns(intrinsicId, baseType);
-        assert(ins != INS_invalid);
+        if (op1 == nullptr)
+        {
+            unreached();
+        }
+        else if (op1->OperIsList())
+        {
+            assert(op2 == nullptr);
+
+            GenTreeArgList* list = op1->AsArgList();
```

Comment:
Is it possible that somebody uses `GenTreeArgList` to encode 2 args form?
```
GT_LIST
+--* arg1
\--* GT_LIST
    +--* arg2
    \--* nullptr
```

Reply:
This should not happen. The list form is supposed to be used only for more than 2 args. And hopefully not for long...

Reply:
Welcome to the new repo @mikedn :smiley: